time_calculation_LLM.py

```diff
@@ -511,45 +542,332 @@ def _expand_gemm_descriptor(gemm: Tuple[int, ...]) -> Tuple[int, int, int, int]:
+        network model sees the same aggregate volume it assumes internally.
+        """
+        if not total_bytes:
+            return 0.0
+
+        if not participants:
+            participants = int(self.tp)
 
-    def _distributed_gemm_forward(self, gemm: Tuple[int, ...], name: str) -> Tuple[float, float]:
+        reduction_time = self.network_model.collective(
+            kind=kind,
+            size_bytes=total_bytes,
+            participants=participants,
+            ib=self.IBTP,
+            ll=self.LLTP,
+            local_bytes=0,
+            debug_label=name or "comm",
+        )
+        return reduction_time
+
+    def parallelism_gemm_forward(self, gemm: Tuple[int, ...], name: str, tensor_type = None) -> Any:
+        parallelism_mode = self.get_parallelism_mode()
+        if parallelism_mode == "tp" or parallelism_mode == "tp-sp":
+            return self._tensor_parallelism_gemm_forward(gemm, name, tensor_type)
+        elif parallelism_mode == "cp":
+            return self._context_parallelism_gemm_forward(gemm, name, tensor_type)
+        elif parallelism_mode == "tp-cp":
+            return self._tensor_context_hybrid_gemm_forward(gemm, name, tensor_type)
+        else:
+            return self.single_gpu_gemm_forward(gemm, name, tensor_type)
+        
+    def parallelism_gemm_backward(self, gemm: Tuple[int, ...], name: str, tensor_type = None) -> Any:
+        parallelism_mode = self.get_parallelism_mode()
+        if parallelism_mode == "tp" or parallelism_mode == "tp-sp":
+            return self._tensor_parallelism_gemm_backward(gemm, name, tensor_type)
+        elif parallelism_mode == "cp":
+            return self._context_parallelism_gemm_backward(gemm, name, tensor_type)
+        elif parallelism_mode == "tp-cp":
+            return self._tensor_context_hybrid_gemm_backward(gemm, name, tensor_type)
+
+        else:
+            return self.single_gpu_gemm_backward(gemm, name, tensor_type)
+    def single_gpu_gemm_forward(self, gemm: Tuple[int, ...], name: str, tensor_type = None) -> Tuple[float, float]:
         batch, m, k, n = self._expand_gemm_descriptor(gemm)
-        if self.t is None or (self.kp1 == 1 and self.kp2 == 1):
-            gemm_time, _, _, mem_access = self.getGEMMTime(m, k, n, name)
-            # dram_access = mem_access[3]
-            # in GB:
-            # dram_access = dram_access / 1e9
-            # print(f"{name} gemm_time: {gemm_time}, mem_access: {dram_access} GB")
-            reduction_time = 0.0
-            gemm_time *= batch # handled internally for CR and RC
-        elif self.t == "CR":
-            gemm_time, reduction_time = self.getDistGEMM_f_kp1(m, k, n, self.kp1, name, batch = batch)
-        elif self.t == "RC":
-            gemm_time, reduction_time = self.getDistGEMM_f_kp2(m, k, n, self.kp1, self.kp2, name, batch = batch)
+        if tensor_type == "attention_score" or tensor_type == "attention_output" :#attention gemm
```

Comment:
we should avoid explicit comparisons with tensor names as much as possible. Do you see how the old code just did *= batch and assumed that non-attn kernels would just have batch = 1 via _expand_gemm_descriptor()? Follow the same philosophy. 
If we hardcode name comparisons the code becomes very hard to follow in the future. If they are "needed" thats ok but try to avoid them.